File: src/main/scala/frentix/event/FFXHREvent.scala
package frentix.event

import scala.collection.mutable
import scala.collection.immutable

/**
 * Created by srosse on 29.09.15.
 */
object FFXHREvent {
  val OFFXHREVENT = "o_ffXHREvent('";

  def apply(link:String):FFXHREvent = {
    val startIndex = link.indexOf(OFFXHREVENT)
    val endIndex = link.indexOf("');", startIndex)
    val cleanedLink = link.substring(startIndex + OFFXHREVENT.length, endIndex)
    val splittedLink: Array[String] = cleanedLink.split(",")
    0 to splittedLink.length - 1 foreach { i =>
        splittedLink(i) = splittedLink(i).stripPrefix("'").stripSuffix("'")
    }
    
    if(splittedLink.length == 9) {
      new FFXHREvent(splittedLink(0), splittedLink(1), splittedLink(2), splittedLink(3), splittedLink(4), splittedLink(8), splittedLink(9), null, null)
    } else {
      new FFXHREvent(splittedLink(0), splittedLink(1), splittedLink(2), splittedLink(3), splittedLink(4), splittedLink(8), splittedLink(9), splittedLink(10), splittedLink(11))
    }
  }
}

/**
 * o_ffXHREvent('ofo_1000006914','ofo_1000006914_dispatchuri','o_fi1000006923','ofo_1000006914_eventval','2',true,true,'cid','selectItem','item','id1399ad03-4981-4240-8a39-dfd9857bf0e8_1:2:1');
 * @param form
 * @param uri
 * @param element
 * @param eventField
 * @param action
 */

class FFXHREvent(form:String, uri:String, element:String, eventField:String, action:String,
                 cmd:String, cmdValue:String, subCmd:String, subCmdValue:String) {

  private val _formName = form
  private val _formId = form.replace("ofo_", "")
  private val _dispatchUri = uri
  private val _elementId = element
  private val _eventFieldId = eventField
  private val _actionId = action
  private val _cmd = cmd
  private val _cmdValue = cmdValue
  private val _subCmd = subCmd
  private val _subCmdValue = subCmdValue

  def formName :String = _formName
  def formId : String = _formId
  def dispatchUri : String = _dispatchUri
  def elementId : String = _elementId
  def eventFieldId : String = _eventFieldId
  def actionId : String = _actionId
  def command : String = _cmd
  def commandValue : String = _cmdValue
  def subCommand : String = _subCmd
  def subCommandValue : String = _subCmdValue

  def formMap(csrfToken:String) : immutable.Map[String,String] = {
    val parameters = collection.mutable.HashMap[String, String]()
    parameters.put("dispatchuri", elementId)
    parameters.put("dispatchevent", actionId)
    if(csrfToken != null) {
      parameters.put("_csrf", csrfToken)
    }
    parameters.put(command, commandValue)
    if(subCommand != null) {
      parameters.put(subCommand, subCommandValue)
    }
    parameters.toMap[String,String]
  }

}

File: src/main/java/org/olat/gatling/event/FFXHREvent.java
/**
 * <a href="https://www.openolat.org">
 * OpenOLAT - Online Learning and Training</a><br>
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License"); <br>
 * you may not use this file except in compliance with the License.<br>
 * You may obtain a copy of the License at the
 * <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache homepage</a>
 * <p>
 * Unless required by applicable law or agreed to in writing,<br>
 * software distributed under the License is distributed on an "AS IS" BASIS, <br>
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. <br>
 * See the License for the specific language governing permissions and <br>
 * limitations under the License.
 * <p>
 * Initial code contributed and copyrighted by<br>
 * frentix GmbH, https://www.frentix.com
 * <p>
 */
package org.olat.gatling.event;

import java.util.HashMap;
import java.util.Map;


/**
 * 
 * Initial date: 29 févr. 2024<br>
 * @author srosse, stephane.rosse@frentix.com, http://www.frentix.com
 *
 */
public record FFXHREvent(String formName,  String formId,
		String dispatchUri, String elementId, String eventFieldId, String actionId,
		String command, String commandValue, String subCommand, String subCommandValue) {
	
	private static final String OFFXHREVENT = "o_ffXHREvent('";
	
	public static final FFXHREvent valueOf(String onclick) {
		if(onclick == null) {
			return null;
		}
		
		int startIndex = onclick.indexOf(OFFXHREVENT);
		int endIndex = onclick.indexOf("');", startIndex);
		String cleanedLink = onclick.substring(startIndex + OFFXHREVENT.length(), endIndex);
		String[] splittedLink = cleanedLink.split(",");
		for(int i=0; i<splittedLink.length; i++) {
			splittedLink[i] = EventHelper.strip(splittedLink[i]);
		}
		
		String optionalSubCommand = null;
		String optionalSubCommandValue = null;
		if(splittedLink.length > 9) {
			optionalSubCommand = splittedLink[10];
			optionalSubCommandValue = splittedLink[11];
		}

		return new FFXHREvent(splittedLink[0], splittedLink[0].replace("ofo_", ""),
				splittedLink[1], splittedLink[2], splittedLink[3], splittedLink[4],
				splittedLink[8], splittedLink[9], optionalSubCommand, optionalSubCommandValue);
	}
	
	public Map<String,String> toMap(String csrfToken) {
		Map<String,String> parameters = new HashMap<>();
		parameters.put("dispatchuri", elementId);
		parameters.put("dispatchevent", actionId);
		if(csrfToken != null && csrfToken.length() > 0) {
			parameters.put("_csrf", csrfToken);
		}
		parameters.put(command, commandValue);
		if(subCommand != null) {
			parameters.put(subCommand, subCommandValue);
		}
		return Map.copyOf(parameters);
	}
}
